lib/chef/resource.rb

```diff
@@ -1169,7 +1236,9 @@ def validate_action(action)
     end
 
     def provider_for_action(action)
-      provider = Chef::ProviderResolver.new(node, self, action).resolve.new(self, run_context)
+      provider_class = self.class.action_provider_class
+      provider_class ||= Chef::ProviderResolver.new(node, self, action).resolve
```

Comment:
We should definitely document that using the `action` keyword in the resource means it will no longer do a provider lookup using the ProviderResolver.  I think this behavior makes sense, though.